blog/2024-07-07-openhab-4-2-release.md

```diff
@@ -0,0 +1,172 @@
+
+X new add-ons, XX bug fixes and XXX enhancements were contributed to the [openHAB add-ons repository](https://github.com/openhab/openhab-addons), which resulted in the addition of XX,xxx lines of code.
+
+The [openHAB web UIs](https://github.com/openhab/openhab-webui) have broken their "personal" best with a whopping XXX pull requests, including XX bug fixes and XX enhancements and resulting in X,xxx lines of code added.
+
+It is worth to mention, that [our other repositories](https://github.com/openhab) have also received numerous contributions (e.g. documentation), but listing them all would probably be too much here.
+Many thanks to all our contributors for driving openHAB forward!
+
+Again, it is time to shout out to our top code contributors per repository (having done the most commits) for this new release:
+
+- Wouter ([@wborn](https://github.com/wborn)) for openHAB core
+- Holger ([@holgerfriedrich](https://github.com/holgerfriedrich)) for openHAB core
+- Jan ([@J-N-K](https://github.com/J-N-K)) for openHAB core
+- Jacob ([@jlaur](https://github.com/jlaur)) for openHAB add-ons
+- Holger ([@holgerfriedrich](https://github.com/holgerfriedrich)) for openHAB add-ons
+- Florian ([@florian-h05](https://github.com/florian-h05)) for the openHAB Main UI
+- Jimmy ([@jimtng](https://github.com/jimtng)) for the openHAB Main UI
+- Laurent ([@lolodomo](https://github.com/lolodomo)) for the openHAB Basic UI
+
+Many of our top committers are also maintainers, which means besides contributing code, they also review pull requests and participate in discussions on GitHub issues, which helps maintain our high level of quality across bindings, UIs and Core.
+At present, we have 36 maintainers across all our repositories.
+Huge thanks to all of them for keeping a high development pace for openHAB 4.2!
+
+# Highlights
+
+As you might have already noticed in the activity statistics, there was much development on openHAB - too much to list everything.
+Please refer to the [**release notes**](https://github.com/openhab/openhab-distro/releases/tag/4.2.0) to check out what's new and noteworthy (e.g. breaking changes).
+
+In the following sections, our maintainers and contributors are introducing some new features you don't want to miss!
+
+## Notification Enhancements
+
+@digitaldan: https://github.com/openhab/openhab-addons/issues/16934
+
+Huge thanks to Dan Cunningham ([@digitaldan](https://github.com/digitaldan)) for doing all the work on openHAB Cloud, the Cloud Connector add-on and the iOS app!
+
+Many thanks to our app maintainers [@weakfl](https://github.com/weakfl), Dan Cunningham ([@digitaldan](https://github.com/digitaldan)), Tim Bert ([@timbms](https://github.com/timbms)) [@mueller-ma](https://github.com/mueller-ma) and Danny Baumann ([@maniac103](https://github.com/maniac103)) for all their work on the apps!
+
+## Core Runtime Enhancements
+
+### Persistence Extensions
+
+@mherwege: https://github.com/openhab/openhab-core/pull/3736
```

Comment:
Persistence Extensions have allowed the different scripting languages to easily persist states and retrieve persisted states or values calculated from these persisted states since a long time.

Since openHAB supports units of measure, persistence services can also return persisted values in a unit of measure.
However, the Persistence Extensions did not consider these units of measure and would return just a decimal value as the result of a calculation.
The Persistence Extension have been enhance to also return the unit of measure.
This allows rules and scripts to now just calculate with QuantityTypes without any implicit or explicit conversion to a DecimalType, while keeping the results consistent if different units of measure are used.
Note that this change may require adapting previous rules.

As persistence now allows forecasts of future states to be stored as well, the Persistence Extensions methods have been enhance to allow persisting future states and retrieving future values.
All persistence retrieval methods now have `last`, `next` and `between` variants of the methods, whereby `last` will look at the period from a time in the past until now, `next` will look from now to a future time and `between` will look between two provided times in past and/or future.

Persistence Extension methods now support working with TimeSeries instead of individual values when persisting or retrieving data.
TimeSeries are series of time/value pairs.
You can construct the series in a rule, e.g. from a TimeSeries retrieved from another persistence service or another item.

Also, there was no way before to retrieve the time of the last state change for an item from persistence.
The existing methods (`previousState()` or `previousState(true)`) would not give the right result if the same value had been persisted multiple times.
A new method `lastChange()` has been added to resolve this.

Scripting languages (rules DSL, javascript, jruby and Blockly) have been enhanced to fully support these new and enhanced Persistence Extensions.

For full documentation of the Persistence Extensions, have a look at [the documentation](https://openhab.org/docs/configuration/persistence.html#persistence-extensions-in-scripts-and-rules).
So much has changed that it is a whortwhile read to update your understanding of openHAB persistence extensions.